File: 131. Palindrome Partitioning/131. Palindrome Partitioning.py
class Solution:
    def partition(self, s: str) -> List[List[str]]:
        def is_palindrome(s):
            return s == s[::-1]

        def dfs(s, path, res):
            if not s:
                res.append(path)
                return
            for i in range(1, len(s) + 1):
                if is_palindrome(s[:i]):
                    dfs(s[i:], path + [s[:i]], res)

        res = []
        dfs(s, [], res)
        return res

File: 131. Palindrome Partitioning/131. Palindrome Partitioning.ts
function partition(s) {
	let res = [];
	dfs(s, [], res);
	return res;
}

function dfs(s, path, res) {
	if (!s) {
		res.push(path);
		return;
	}
	for (let i = 1; i <= s.length; i++) {
		let sub = s.slice(0, i);
		if (isPalindrome(sub)) {
			dfs(s.slice(i), path.concat(sub), res);
		}
	}
}

function isPalindrome(s) {
	let left = 0;
	let right = s.length - 1;
	while (left < right) {
		if (s[left] !== s[right]) {
			return false;
		}
		left++;
		right--;
	}
	return true;
}
